README.md
# @volebo/express

```yaml
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

[![Build Status](https://travis-ci.org/volebo/express.svg?branch=master)](https://travis-ci.org/volebo/express)
[![codecov](https://codecov.io/gh/volebo/express/branch/master/graph/badge.svg)](https://codecov.io/gh/volebo/express)
[![bitHound Overall Score](https://www.bithound.io/github/volebo/express/badges/score.svg)](https://www.bithound.io/github/volebo/express)
[![bitHound Dependencies](https://www.bithound.io/github/volebo/express/badges/dependencies.svg)](https://www.bithound.io/github/volebo/express/master/dependencies/npm)
[![npm version](https://img.shields.io/npm/v/@volebo/express.svg)](https://www.npmjs.com/package/@volebo/express)
[![npm downloads](https://img.shields.io/npm/dm/@volebo/express.svg)](https://www.npmjs.com/package/@volebo/express)

https://codeclimate.com/github/volebo/volebo-express

Express JS server for Volebo.Net

## Usage

First, use this project as a dependency for your project.

Since the codebase isn't mature (at any level), it is better to use
`npm link @volebo/express` than `npm install`.

Then, just follow this sample:

```javascript
'use strict'

// require the module:
const vbexpress       = require('@volebo/express')

const configPath = 'server/config.yaml'
const overrideOpts = {
	'debug': true
}

// create an instance with appropriate options:
const app = vbexpress(configPath, overrideOpts)

// Add routes (no lang-handler):
const routes = require('./routes')
app.use('/', routes)

// Add routes with language-culture detection:
const routesL10N = require('./routes2')
app.lang.use('/', routesL10N)

// start handling
app.start()
```

## Additional properties

#### `app.config`

Custom configuration of the server. It contains our own crutches and fixes.

#### `app.hbs`

Handlebars engine with registered [HBS-helpers](http://assemble.io/helpers/)
by [Assemble]-team

#### `app.lang`

Express router, created by the [`express-mw-lang`][express-mw-lang] module.

#### `app.log`

Bunyan [logger][npm-bunyan].

### Request and Response

This module extends standard `req` and `res` with additional fields. Also, this
module is sensitive to several fields of the `req` and `res`.

#### `res.helpers`

This object will be used as **per-request** HBS-helpers (see
[this documentation][express-hbs-helpers] for additional details).

The main purpose: set **locale** for handling current request with client's
locale settings.

#### `res.locals.lang`

The extensions for res.locals, generated by the [express-mw-lang] module.

#### `res.locals.user`

Information about logged-in user. User is authenticated (by [PassportJS]) and
partially filtered (a few fields are accessible).

#### `req.lang`

See [`res.locals.lang`](#res-locals-lang)

#### `req.forwardedSecure`

Describes, whether the front end server currently handles HTTPS query (after
proxying it becomes HTTP).

## Credits

All thanks and praises go to:

* [ExpressJS](http://expressjs.com) and many contributors of the
    [ExpressJS org](https://github.com/expressjs)
* [Express Handlebars](https://github.com/ericf/express-handlebars)
* [PassportJS] and Jared Hanson
* [Bunyan Logger][npm-bunyan]
* [MessageFormat] for their implementation of ICU standard (I18N/L10N)
* [Sentry] - awesome error collector

## License

Please, follow the link: [LICENSE](https://github.com/volebo/volebo-express/raw/master/LICENSE)

[express-mw-lang]: https://www.npmjs.com/package/express-mw-lang
[express-hbs-helpers]: https://www.npmjs.com/package/express-handlebars#helpers-1
[npm-bunyan]: https://www.npmjs.com/package/bunyan

[PassportJS]: http://passportjs.org/
[Assemble]: http://assemble.io
[MessageFormat]: https://messageformat.github.io
[Sentry]: https://sentry.io


CHANGELOG.md
# CHANGELOG

```yaml
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

This is the history of changes of the `@volebo/express` package

> This file should be filled by maintainers, using pull requests
> Please, follow this [guide](http://keepachangelog.com/en/0.3.0/)

## unpublished // ??

## 0.8.4 // 2018-04-03

* update dependencies

## 0.8.0 // 2017-10-31

* optional **Sentry** integration (gh #2)
* allow to configure **Helmet**
* use [messageformat.js](https://github.com/messageformat/messageformat.js) for i18n
* use [Assemble](http://assemble.io/helpers/) as HBS-helpers collection (gh #17)

* move several functions to `/loaders`
* move default configuration to YAML file (easier to maintain and copy)
* rename tests (introduce `@volebo/mocha-helpers`)
* other refactoring stuff

#### Issues

* see #17 - HBS helpers
* see #2 - error handling

## 0.7.6 // 2017-10-27

* **rename to `@volebo/express`** (gh #10)

* see #10

## 0.7.5 // 2017-09-15

* proper way to auto-load `@volebo/data`
* strict typed Errors. First class `NotFoundError`

* fix volebo/express#22 - logdir handling

## 0.7.3 // 2017-08-15

* data: optional model loading

## 0.7.2 // 2017-08-15

* log: write to graylog (UDP, not-configurable yet); add new property: `app.log`
* HBS: rename helper __ => t, and make it alive (still a dirty implementation)
* auth: load `passport` user to `res.locals.user`
* security: set default cookie domain in config

## 0.7.1 // 2017-08-07

* new config object `Config`. Access to configuration only with `app.config.get('path')`
* config files must have `volebo` as root node
* easy to load config from the environment, using variables like `volebo_model_enabled`
* add `log` dir to git, because FYTIW (new bug: gh #22)
* set default value for `session.domains`: `['volebo.net']`
* accurate model create/dispose
* use [`express-flash`](https://www.npmjs.com/package/express-flash) (gh #24)

* well-tune `istanbul` (coverage tool)

## 0.6.1 // 2017-08-06

This record includes changes from previos releases

* remove getnodeenv (app should not care about env)
* use bunyan logger as common logger
* CONFIG: renaming: `model.db` => `db.connection`
* new field for app: `app.passport`
* I18N: `app.lang.loadTranslation` - for loading translations to App memory
* HBS: move views to new folder
* HBS: reorganize helpers (init through function)
* ERROR handling: fixed bug with unaccessible layout for error-rendering
* migrate to `bunyan` (instead of `morgan`) for logging (gh #2)

* see #2

## 0.3.1

* autopublish from travis

## 0.3.0

* new eslint-rules
* simplified get-node-env scripts (removed `lodash` dep)
* change copyright years
* stub for helpers (see #17)

## 0.2.9 // 2016-12-10

* use helmet
* better main contructor
* better NODE_ENV handling
* set `trust proxy`
* reorganize tests

* fix #12
* fix #9
* fix #16

## 0.2.8 // 2016-12-03

* set timezone to **UTC** for `knex` in the `config`
* more docs

* fix #15

## 0.2.7 // 2016-11-04

* Config: pretty JSON for dev-env
* disable `x-powered-by`
* `config.db` moved to `config.model.db`
* autoload model (`volebo-data`)
* listening on local path (not host:port), handle existing socket correctly


src/server.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################

ExpressJS for volebo.net

Copyright (C) 2016-2018 Volebo <dev@volebo.net>
Copyright (C) 2016-2018 Maksim Koryukov <maxkoryukov@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

'use strict'

// Loading polyfill for intl
// required, as soon we want to use several locales (not only EN)
if (global.Intl) {
	const IntlPolyfill = require('intl')
	Intl.NumberFormat   = IntlPolyfill.NumberFormat
	Intl.DateTimeFormat = IntlPolyfill.DateTimeFormat
} else {
	global.Intl = require('intl')
}

const path            = require('path')

const debug           = require('debug')('volebo:express:server')
const express         = require('express')
const helmet          = require('helmet')

const bodyParser      = require('body-parser')
const handlebars      = require('express-handlebars')

const session         = require('express-session')
const flash           = require('express-flash')

const _               = require('lodash')
const moment          = require('moment')

const raven           = require('raven')

const langGen         = require('express-mw-lang')

const logger          = require('./logger')

const errors             = require('./errors')
const loadConfiguration  = require('./loaders/configuration')
const readL10n           = require('./loaders/l10n')
const loadModelModule    = require('./loaders/safe-model-module')
const registerHbsHelpers = require('./hbs-helpers')


const log = logger.getLogger('volebo.express.server')


debug('initializing')

const VoleboModel       = loadModelModule(log)


function main(configPath, overrideOptions) {
	const app = express()

	/*
	========================================================
	CONFIG

	Custom config, allows to put dirty crutches
	without breaking core settings of the express
	========================================================
	*/
	app.config = loadConfiguration(configPath, overrideOptions)
	app.log = log

	// securing with HTTP-headers
	app.use(helmet(app.config.get('security.helmet', {})))

	/*
	========================================================
	SENTRY stuff
	========================================================
	*/
	if (app.config.get('sentry.enabled')) {
		// Must configure Raven before doing anything else with it
		const _dsn = app.config.get('sentry.dsn')
		const _opts = app.config.get('sentry.options') || {}

		const _release = app.config.get('package.version')
		if (!_.isNil(_release)) {
			_opts['release'] = _release
		}

		debug('Sentry options', _opts)
		raven.config(_dsn, _opts).install()

		// The request handler must be the first middleware on the app
		app.use(raven.requestHandler())
	}

	/*
	========================================================
	AUTOLOAD MODEL
	========================================================
	*/
	if (app.config.get('model.enabled')) {
		app.model = new VoleboModel(app.config.get('model'))

		app.log.info(`app.model attached, type = [${typeof app.model}]`)
	}
	debug('app.model attached:', !!app.model)

	/*
	========================================================
	COMMON MIDDLEWARES
	========================================================
	*/

	app.use(logger.getWwwLogger('volebo.express.server.www'))
	app.use(logger.getWwwErrorLogger('volebo.express.server.www'))

	app.use(bodyParser.json())
	app.use(bodyParser.urlencoded({ extended: false }))

	app.set('trust proxy', app.config.get('proxy.list'))

	/*
	========================================================
	STATIC
	========================================================
	*/

	if (app.config.get('debug.staticPath')) {
		//let staticPath = path.join(__dirname, app.config.debug.staticPath);
		const staticPath = app.config.get('debug.staticPath')

		debug('Use static path', staticPath)
		app.use(express.static(staticPath))
	}

	/*
	========================================================
	JSON
	========================================================
	*/
	if (app.config.debug) {
		app.set('json spaces', 4)
	}

	/*
	========================================================
	HTTP->HTTPS redirect (working behind NGINX)
	========================================================
	*/
	app.use((req, _unused_res, next) => {
		req.forwardedSecure = _.lowerCase(req.headers["x-forwarded-proto"]) === "https"
		next()
	})

	/*
	========================================================
	SESSIONS
	========================================================
	*/

	if (app.config.get('session.enabled')) {

		if (-1 === app.config.get('session.secret')) {
			log.error('Seems like you forgot to set session.secret in your config')
		}

		const session_config = {
			name: app.config.get('session.name', 'sessionId'),
			secret: app.config.get('session.secret'),
			resave: false,
			saveUninitialized: false,

			// DOC: https://www.npmjs.com/package/express-session#cookie
			cookie: {
				//expires
				maxAge: 1000*60*30, // ms, 30 minutes now
				httpOnly: true,

				secure : app.config.get('session.secure'),
			},
			/*genid: function(req) { return genuuid() // use UUIDs for session IDs  },*/
		}

		const session_domains = app.config.get('session.domains')
		if (_.isArray(session_domains) && session_domains.length > 0) {
			session_config.cookie.domain = session_domains.join(',')
		}

		app.use(session(session_config))
	}

	if (app.config.get('flash.enabled')) {
		app.use(flash())
	}

	/*
	========================================================
	PASSPORT
	========================================================
	*/

	app.passport = null

	if (app.config.get('auth.enabled')) {

		app.passport = require('passport')
		app.use(app.passport.initialize())

		if (app.config.get('auth.session')){
			app.use(app.passport.session())

			app.passport.serializeUser(function(user, done) {
				// TODO: improve
				debug('serializing', user)
				done(null, user)
			})

			app.passport.deserializeUser(function(obj, done) {
				// TODO: improve
				debug('deserializing', obj)
				done(null, obj)
			})
		}

		app.use(function(req, res, next) {
			let u = null
			if (req.user) {
				u = _(req.user)
					.pick(req.user, [
						'username',
					])
					.mapKeys((_, key) => {
						return {
							'username': 'username'
						}[key]
					})
					.value()
			}
			debug('append user info to req.locals', u)
			res.locals.user = u

			return next()
		})
	}

	/*
	========================================================
	Handler for lang detection and other stuff..
	========================================================
	*/

	const langmw = langGen({
		defaultLanguage: 'en',
		availableLanguages: ['en', 'ru', 'zh'],
		onLangDetected: function(lang_code, _unused_req, res) {

			moment.locale(lang_code)

			// TODO: move this to the separated middleware "attach helpers"

			// REF #6: We could use `defaultOptions`, attached to the
			// `res` object. This approach could make the code
			// more readable.
			// See also: _renderTemplate hook
			// https://github.com/ericf/express-handlebars#_rendertemplatetemplate-context-options

			if (!res._renderOriginal) {
				res._renderOriginal = res.render
			}
			res.render = function overloadedRender(){

				const nargs = Array.prototype.slice.call(arguments)
				const opt = _.get(nargs, '[1]', {})

				if (res.helpers) {
					debug('Going to append per-request HBS helpers')

					// check, whether some helpers are already attached:
					if(opt.helpers) {
						_.assign(opt.helpers, res.helpers)
					} else {
						_.set(opt, 'helpers', res.helpers)
					}
				}

				// set locale for HBS-intl formatter (format dates, numbers and messages)
				_.set(opt, 'data.intl.locales', lang_code)

				nargs[1] = opt
				return res._renderOriginal.apply(this, nargs)
			}
		}
	})
	app.lang = langmw
	app.l10n = {}

	const _availableLangCodes = _.map(langmw.available, 'code')
	app.lang.loadL10n = function app_lang_loadL10n(folder) {
		const _newL10n = readL10n(folder, _availableLangCodes)
		_.assign(app.l10n, _newL10n)
	}

	// BUG: gh #4 fix this CWD shit
	app.lang.loadL10n(path.join(process.cwd(), 'l10n'))

	langmw.esu(app)

	/*
	========================================================
	VIEW ENGINE SETUP
	========================================================
	*/
	const hbs  = handlebars.create({
		layoutsDir: path.join(__dirname, 'views', 'layouts'),
		partialsDir: path.join(__dirname, 'views', 'partials'),		// TODO : #13 use NAMESPACES
		defaultLayout: 'default',
		extname: '.hbs'
	})
	app.set('views', path.join(__dirname, 'views'))
	app.hbs = hbs
	app.hbs.helpers = registerHbsHelpers(app)

	app.engine('hbs', hbs.engine)
	app.set('view engine', 'hbs')

	app._onStarting = function() {
		/*
		====================================
		NOT FOUND HANDLER
		catch 404 and forward to error handler
		TODO: we could
		====================================
		*/
		app.use(function(req, _unused_res, next) {
			// TODO : pass required info to the error, such as URL, params...
			const err = new errors.NotFoundError(req)

			return next(err)
		})

		/*
		====================================
		ERROR HANDLERS
		TODO: fix #2 - robust error handler
		====================================
		*/

		const errorViewPath = path.join(__dirname, 'views', 'error.hbs')
		const errorLayoutPath = path.join(__dirname, 'views', 'layouts', 'default.hbs')

		if (app.config.get('sentry.enabled')) {
			// The error handler must be before any other error middleware
			debug('Sentry: connect error handler')
			app.use(raven.errorHandler())
		}

		app.use(function global_error_dev(err, _unused_req, res, next) {
			res.status(err.status || 500)

			/*
			// When sentry is enabled, additional info is accessble
			// res.end(res.sentry + '\n');
			*/

			// development error handler
			// will print stacktrace
			// no stacktraces leaked to user
			const pubStack = app.config.get('debug.renderStack')
				? err.stack
				: null

			app.log.error(err)

			return res.render(errorViewPath, {
				message: err.message,
				stack: pubStack,
				status: err.status,
				isDevMode: app.config.debug,

				layout: errorLayoutPath,
			})
		})
	}

	return app
}

exports = module.exports = main
